src/os_common/os_alloc.c

```diff
@@ -147,6 +147,61 @@ __realloc_func(WT_SESSION_IMPL *session, size_t *bytes_allocated_ret, size_t byt
     return (0);
 }
 
+/*
+ * __realloc_malloc_func --
+ *     ANSI realloc function.
+ */
+static int
+__realloc_malloc_func(WT_SESSION_IMPL *session, size_t *bytes_allocated_ret,
+  size_t bytes_to_allocate, bool clear_memory, void *retp)
+{
+    size_t bytes_allocated;
+    void *p, *p1;
+    /*
+     * !!!
+     * This function MUST handle a NULL WT_SESSION_IMPL handle.
+     *
+     * Sometimes we're allocating memory and we don't care about the
+     * final length -- bytes_allocated_ret may be NULL.
+     */
+    p = *(void **)retp;
+    bytes_allocated = (bytes_allocated_ret == NULL) ? 0 : *bytes_allocated_ret;
+    WT_ASSERT(session,
+      (p == NULL && bytes_allocated == 0) ||
+        (p != NULL && (bytes_allocated_ret == NULL || bytes_allocated != 0)));
+    WT_ASSERT(session, bytes_to_allocate != 0);
+    WT_ASSERT(session, bytes_allocated < bytes_to_allocate);
+
+    if (session != NULL) {
+        if (p == NULL)
+            WT_STAT_CONN_INCR(session, memory_allocation);
+        else
+            WT_STAT_CONN_INCR(session, memory_grow);
+    }
+
+    if ((p1 = malloc(bytes_to_allocate)) == NULL)
+        WT_RET_MSG(session, __wt_errno(), "memory allocation of %" WT_SIZET_FMT " bytes failed",
+          bytes_to_allocate);
+
+    memcpy(p, p1, *bytes_allocated_ret);
+    memset((uint8_t *)p, 0, bytes_allocated);
```

Comment:
Instead of setting memory to `0`, I think we should have `WT_DEBUG_BYTE`.  This will make references to the old memory more obvious.